src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -52,33 +54,69 @@ public override string Description
         /// <returns></returns>
         protected override bool Matches(IEnumerable actual)
         {
-            return TryFastAlgorithm(actual) ?? OriginalAlgorithm(actual);
+            var nonUniqueItems = TryFastAlgorithm(actual) ?? OriginalAlgorithm(actual);
+            return nonUniqueItems.Count == 0;
+        }
+
+        /// <inheritdoc />
+        public override ConstraintResult ApplyTo<TActual>(TActual actual)
+        {
+            IEnumerable enumerable = ConstraintUtils.RequireActual<IEnumerable>(actual, nameof(actual));
+
+            var nonUniqueItems = TryFastAlgorithm(enumerable) ?? OriginalAlgorithm(enumerable);
+
+            return new UniqueItemsContstraintResult(this, actual, nonUniqueItems);
         }
 
-        private bool OriginalAlgorithm(IEnumerable actual)
+        private ICollection OriginalAlgorithm(IEnumerable actual)
```

Comment:
I haven't tested this - having some computer problems 😄 - but couldn't this be simplified into ?
```csharp
        private ICollection OriginalAlgorithm(IEnumerable actual)
        {
            var processedItems = new List<object>();
            var nonUniques = new List<object>();

            foreach (object o1 in actual)
            {
                var isNonUnique = false;
                var knownNonUnique = false;

                foreach (object o2 in processedItems)
                {
                    if (ItemsEqual(o1, o2))
                    {
                        isNonUnique = true;
                        knownNonUnique = nonUniques.Where(o2 => ItemsEqual(o1, o2)).Any();
                        break;
                    }
                }

                if (!isNonUnique)
                    processedItems.Add(o1);
                else if (!knownNonUnique)
                    nonUniques.Add(o1);
            }

            return nonUniques;
        }
```

Reply:
That worked, thanks!